### PR TITLE
Collect contained ObjectIDs during deserialization

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Required for reference counting serialized ObjectIDs. Will be hooked up to reference counting logic in a follow-up PR.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -287,7 +287,13 @@ cdef void prepare_args(
         else:
             serialized_arg = worker.get_serialization_context().serialize(arg)
             size = serialized_arg.total_bytes
-            if <int64_t>size <= put_threshold:
+
+            # TODO(edoakes): any objects containing ObjectIDs are spilled to
+            # plasma here. This is inefficient for small objects, but inlined
+            # arguments aren't associated ObjectIDs right now so this is a
+            # simple fix for reference counting purposes.
```
Comment (stephanie-wang) [inline python/ray/_raylet.pyx]: Hmm why is this necessary? I think we only need to do this for object IDs that were contained in another ObjectID (which we did [here](https://github.com/ray-project/ray/blob/master/src/ray/core_worker/core_worker.cc#L333)), right?

File: python/ray/serialization.py [hunk 2]
```diff
@@ -202,12 +215,22 @@ def _deserialize_object(self, data, metadata, object_id):
...
+                        obj = pickle.loads(in_band, buffers=buffers)
                     else:
-                        return pickle.loads(in_band)
+                        obj = pickle.loads(in_band)
                 # cloudpickle does not provide error types
                 except pickle.pickle.PicklingError:
                     raise DeserializationError()
+
+                if object_id.is_nil():
+                    assert len(self.get_and_clear_contained_object_ids()) == 0
```
Comment (stephanie-wang) [inline python/ray/serialization.py]: When could this happen?

File: python/ray/serialization.py [hunk 3]
```diff
@@ -147,6 +154,7 @@ def id_deserializer(serialized_obj):
             return serialized_obj[0](*serialized_obj[1])
 
         def object_id_serializer(obj):
+            self.add_contained_object_id(obj)
```
Comment (stephanie-wang) [inline python/ray/serialization.py]: Should this be defined somewhere?

File: python/ray/serialization.py [hunk 4]
```diff
@@ -202,12 +215,22 @@ def _deserialize_object(self, data, metadata, object_id):
...
                 except pickle.pickle.PicklingError:
                     raise DeserializationError()
+
+                if object_id.is_nil():
+                    assert len(self.get_and_clear_contained_object_ids()) == 0
+                else:
+                    worker = ray.worker.global_worker
+                    worker.core_worker.add_contained_object_ids(
+                        object_id,
+                        self.get_and_clear_contained_object_ids(),
```
Comment (stephanie-wang) [inline python/ray/serialization.py]: Should this be defined somewhere?

**Review by stephanie-wang — COMMENTED** [review]
Looks good, but I was a bit confused about where the `add_contained_object_id` method gets defined. Is that pushed?

**Comment by edoakes** [comment]
Ah yeah sorry that's in the PR that this is built on top of and there was a bad merge when Philipp merged the commit that removes pyarrow serialization. Fixed now.

File: python/ray/serialization.py (hunk 2 above)
Comment (edoakes) [reply python/ray/serialization.py]: This happens when arguments are inlined. Added a comment.

File: python/ray/_raylet.pyx (hunk 1 above)
Comment (stephanie-wang) [reply python/ray/_raylet.pyx]: Ah nvm, I understand now. Yeah, it is a bit unfortunate, but I guess later we can fix it by assigning inlined objects ObjectIDs as well. Actually shouldn't be too hard since we have the in-memory store.

**Review by stephanie-wang — APPROVED** [review]

File: python/ray/_raylet.pyx (hunk 1 above)
Comment (edoakes) [reply python/ray/_raylet.pyx]: Yep, I think the best solution is to just `Put()` everything and put in in-memory store/plasma based on size in the core worker.
